# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [3]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [4]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    source_id_text = list()
    
    for sentence in source_text.split('\n'):
        sent_id = list()
        for word in sentence.split(' '):
            try:
                sent_id.append(source_vocab_to_int[word])
            except:
                sent_id.append(source_vocab_to_int['<UNK>'])
        source_id_text.append(sent_id)        
        
        
    target_id_text = list()
    
    for sentence in target_text.split('\n'):
        sent_id = list()
        for word in sentence.split(' '):
            try:
                sent_id.append(target_vocab_to_int[word])
            except:
                sent_id.append(target_vocab_to_int['<UNK>'])
        sent_id.append(target_vocab_to_int['<EOS>'])
        target_id_text.append(sent_id)        
        
    
    return  source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.2.1


C:\Users\Emil\AppData\Local\Continuum\Anaconda2\envs\py35\lib\site-packages\ipykernel\__main__.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [8]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    
    # TODO: Implement Function
    
    input = tf.placeholder(tf.int32, [None, None], name = 'input')
    targets = tf.placeholder(tf.int32, [None, None], name = 'targets')
    learnrate = tf.placeholder(tf.float32, name = 'learnrate')
    keep_prob = tf.placeholder(tf.float32, name = 'keep_prob')
    target_sequence_length = tf.placeholder(tf.int32, [None], name = 'target_sequence_length')
    max_target_len =  tf.reduce_max(target_sequence_length)
    source_sequence_length = tf.placeholder(tf.int32, [None], name = 'source_sequence_length')
    
    
    return input, targets, learnrate, keep_prob,  target_sequence_length, max_target_len, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "C:\\Users\\Emil\\AppData\\Local\\Continuum\\Anaconda2\\envs\\py35\\lib\\runpy.py", line 184, in _run_module_as_main\n    "__main__", mod_spec)', 'File "C:\\Users\\Emil\\AppData\\Local\\Continuum\\Anaconda2\\envs\\py35\\lib\\runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "C:\\Users\\Emil\\AppData\\Local\\Continuum\\Anaconda2\\envs\\py35\\lib\\site-packages\\ipykernel\\__main__.py", line 3, in <module>\n    app.launch_new_instance()', 'File "C:\\Users\\Emil\\AppData\\Local\\Continuum\\Anaconda2\\envs\\py35\\lib\\site-packages\\traitlets\\config\\application.py", line 658, in launch_instance\n    app.start()', 'File "C:\\Users\\Emil\\AppData\\Local\\Continuum\\Anaconda2\\envs\\p

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [9]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    ending = tf.strided_slice(target_data, [0,0], [batch_size, -1], [1,1])
    decoder_input = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    
    return decoder_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [12]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    emb_layer = tf.contrib.layers.embed_sequence( rnn_inputs, 
                                        vocab_size=source_vocab_size,
                                        embed_dim=encoding_embedding_size)
    
    def lstm_cell():
        lstm_cell = tf.contrib.rnn.LSTMCell(rnn_size, initializer = tf.random_uniform_initializer(-0.1 , 0.1, seed = 2))
        return tf.contrib.rnn.DropoutWrapper(lstm_cell, input_keep_prob=keep_prob)
        
        
    stacked_lstm = tf.contrib.rnn.MultiRNNCell([lstm_cell() for _ in range(num_layers)])

    encoder_output, encoder_state = tf.nn.dynamic_rnn(stacked_lstm, emb_layer, sequence_length=source_sequence_length, dtype=tf.float32)

    return encoder_output, encoder_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [13]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    
    train_helper = tf.contrib.seq2seq.TrainingHelper(inputs = dec_embed_input, 
                                                    sequence_length = target_sequence_length,
                                                    time_major = False)
    
    basic_decoder = tf.contrib.seq2seq.BasicDecoder( dec_cell, train_helper, encoder_state, output_layer)
    
    final_decode_output, _ , _ = tf.contrib.seq2seq.dynamic_decode(basic_decoder, impute_finished=True, maximum_iterations= max_summary_length)
    
    return final_decode_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [14]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    
    
    start = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name = 'start')
    
    
    
    infer_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens=start, end_token=end_of_sequence_id)
    
    infer_decoder = tf.contrib.seq2seq.BasicDecoder( dec_cell, infer_helper, encoder_state, output_layer)
    
    infer_decode_output, _ , _ = tf.contrib.seq2seq.dynamic_decode(infer_decoder, impute_finished=True, maximum_iterations= max_target_sequence_length)
    
    return infer_decode_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [17]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    decoding_emb = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    
    decoding_emb_input = tf.nn.embedding_lookup(decoding_emb, dec_input)
    
    def lstm_cell(rnn__size):
        lstm_cell = tf.contrib.rnn.LSTMCell(rnn_size, initializer = tf.random_uniform_initializer(-0.1 , 0.1, seed = 2))
        return tf.contrib.rnn.DropoutWrapper(lstm_cell, input_keep_prob=keep_prob)
    
    decoding_cell = tf.contrib.rnn.MultiRNNCell([lstm_cell(rnn_size) for _ in range(num_layers)])
    
    output = Dense(target_vocab_size,
                  kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev = 0.1))
    
    
    train_logits = decoding_layer_train(encoder_state, decoding_cell, decoding_emb_input,
                                       target_sequence_length, max_target_sequence_length,
                                       output, keep_prob)
    
    infer_logits = decoding_layer_infer(encoder_state, decoding_cell, decoding_emb,
                                       target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'],
                                       max_target_sequence_length, target_vocab_size,
                                       output, batch_size, keep_prob)
    
    
    
    return train_logits, infer_logits



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [23]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    
    encoder_emb_input = tf.contrib.layers.embed_sequence(input_data, source_vocab_size, enc_embedding_size)
    
    encoder_output , encoder_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob,
                                       source_sequence_length, source_vocab_size,
                                       enc_embedding_size)
    
    decoder_input = process_decoder_input(target_data, target_vocab_to_int, batch_size) 
    
    decoder_embedings = tf.Variable(tf.random_uniform([target_vocab_size, dec_embedding_size]))
    
    
    train_dec_logits, infer_dec_logits = decoding_layer(decoder_input, encoder_state, target_sequence_length,
                                                        max_target_sentence_length, rnn_size,
                                                        num_layers, target_vocab_to_int,
                                                        target_vocab_size, batch_size, keep_prob,
                                                        dec_embedding_size)
    
    
    
    return train_dec_logits, infer_dec_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [45]:
# Number of Epochs
epochs = 10
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.5
display_step = 10

### Build the Graph
Build the graph using the neural network you implemented.

In [46]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [47]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [48]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   10/1077 - Train Accuracy: 0.2928, Validation Accuracy: 0.3896, Loss: 3.6535
Epoch   0 Batch   20/1077 - Train Accuracy: 0.3699, Validation Accuracy: 0.4158, Loss: 2.8704
Epoch   0 Batch   30/1077 - Train Accuracy: 0.3879, Validation Accuracy: 0.4556, Loss: 2.6506
Epoch   0 Batch   40/1077 - Train Accuracy: 0.4254, Validation Accuracy: 0.4808, Loss: 2.4777
Epoch   0 Batch   50/1077 - Train Accuracy: 0.4211, Validation Accuracy: 0.4929, Loss: 2.3715
Epoch   0 Batch   60/1077 - Train Accuracy: 0.4695, Validation Accuracy: 0.5107, Loss: 2.1308
Epoch   0 Batch   70/1077 - Train Accuracy: 0.4478, Validation Accuracy: 0.5245, Loss: 2.0761
Epoch   0 Batch   80/1077 - Train Accuracy: 0.4188, Validation Accuracy: 0.4854, Loss: 1.8854
Epoch   0 Batch   90/1077 - Train Accuracy: 0.4129, Validation Accuracy: 0.4670, Loss: 1.8058
Epoch   0 Batch  100/1077 - Train Accuracy: 0.4477, Validation Accuracy: 0.4890, Loss: 1.6632
Epoch   0 Batch  110/1077 - Train Accuracy: 0.3992, Validati

Epoch   0 Batch  890/1077 - Train Accuracy: 0.7909, Validation Accuracy: 0.7255, Loss: 0.3563
Epoch   0 Batch  900/1077 - Train Accuracy: 0.7348, Validation Accuracy: 0.7326, Loss: 0.3738
Epoch   0 Batch  910/1077 - Train Accuracy: 0.7206, Validation Accuracy: 0.7205, Loss: 0.3590
Epoch   0 Batch  920/1077 - Train Accuracy: 0.7672, Validation Accuracy: 0.7305, Loss: 0.3691
Epoch   0 Batch  930/1077 - Train Accuracy: 0.7586, Validation Accuracy: 0.7500, Loss: 0.3445
Epoch   0 Batch  940/1077 - Train Accuracy: 0.7637, Validation Accuracy: 0.7433, Loss: 0.3488
Epoch   0 Batch  950/1077 - Train Accuracy: 0.7705, Validation Accuracy: 0.7390, Loss: 0.3145
Epoch   0 Batch  960/1077 - Train Accuracy: 0.7690, Validation Accuracy: 0.7461, Loss: 0.3284
Epoch   0 Batch  970/1077 - Train Accuracy: 0.7969, Validation Accuracy: 0.7773, Loss: 0.3384
Epoch   0 Batch  980/1077 - Train Accuracy: 0.8168, Validation Accuracy: 0.7820, Loss: 0.3241
Epoch   0 Batch  990/1077 - Train Accuracy: 0.7759, Validati

Epoch   1 Batch  700/1077 - Train Accuracy: 0.9172, Validation Accuracy: 0.9020, Loss: 0.0697
Epoch   1 Batch  710/1077 - Train Accuracy: 0.9199, Validation Accuracy: 0.9165, Loss: 0.0673
Epoch   1 Batch  720/1077 - Train Accuracy: 0.9145, Validation Accuracy: 0.9180, Loss: 0.0822
Epoch   1 Batch  730/1077 - Train Accuracy: 0.9352, Validation Accuracy: 0.9272, Loss: 0.0863
Epoch   1 Batch  740/1077 - Train Accuracy: 0.9363, Validation Accuracy: 0.9126, Loss: 0.0623
Epoch   1 Batch  750/1077 - Train Accuracy: 0.9316, Validation Accuracy: 0.9226, Loss: 0.0695
Epoch   1 Batch  760/1077 - Train Accuracy: 0.9375, Validation Accuracy: 0.8988, Loss: 0.0876
Epoch   1 Batch  770/1077 - Train Accuracy: 0.9252, Validation Accuracy: 0.9091, Loss: 0.0682
Epoch   1 Batch  780/1077 - Train Accuracy: 0.8934, Validation Accuracy: 0.9187, Loss: 0.0950
Epoch   1 Batch  790/1077 - Train Accuracy: 0.8844, Validation Accuracy: 0.9400, Loss: 0.0731
Epoch   1 Batch  800/1077 - Train Accuracy: 0.9336, Validati

Epoch   2 Batch  510/1077 - Train Accuracy: 0.9379, Validation Accuracy: 0.9411, Loss: 0.0470
Epoch   2 Batch  520/1077 - Train Accuracy: 0.9807, Validation Accuracy: 0.9428, Loss: 0.0380
Epoch   2 Batch  530/1077 - Train Accuracy: 0.9437, Validation Accuracy: 0.9471, Loss: 0.0435
Epoch   2 Batch  540/1077 - Train Accuracy: 0.9551, Validation Accuracy: 0.9450, Loss: 0.0342
Epoch   2 Batch  550/1077 - Train Accuracy: 0.9531, Validation Accuracy: 0.9386, Loss: 0.0367
Epoch   2 Batch  560/1077 - Train Accuracy: 0.9434, Validation Accuracy: 0.9528, Loss: 0.0368
Epoch   2 Batch  570/1077 - Train Accuracy: 0.9498, Validation Accuracy: 0.9506, Loss: 0.0571
Epoch   2 Batch  580/1077 - Train Accuracy: 0.9654, Validation Accuracy: 0.9499, Loss: 0.0310
Epoch   2 Batch  590/1077 - Train Accuracy: 0.9186, Validation Accuracy: 0.9609, Loss: 0.0474
Epoch   2 Batch  600/1077 - Train Accuracy: 0.9516, Validation Accuracy: 0.9364, Loss: 0.0481
Epoch   2 Batch  610/1077 - Train Accuracy: 0.9289, Validati

Epoch   3 Batch  320/1077 - Train Accuracy: 0.9695, Validation Accuracy: 0.9474, Loss: 0.0299
Epoch   3 Batch  330/1077 - Train Accuracy: 0.9508, Validation Accuracy: 0.9613, Loss: 0.0305
Epoch   3 Batch  340/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9503, Loss: 0.0313
Epoch   3 Batch  350/1077 - Train Accuracy: 0.9688, Validation Accuracy: 0.9421, Loss: 0.0260
Epoch   3 Batch  360/1077 - Train Accuracy: 0.9672, Validation Accuracy: 0.9485, Loss: 0.0238
Epoch   3 Batch  370/1077 - Train Accuracy: 0.9758, Validation Accuracy: 0.9563, Loss: 0.0322
Epoch   3 Batch  380/1077 - Train Accuracy: 0.9840, Validation Accuracy: 0.9524, Loss: 0.0239
Epoch   3 Batch  390/1077 - Train Accuracy: 0.9473, Validation Accuracy: 0.9631, Loss: 0.0355
Epoch   3 Batch  400/1077 - Train Accuracy: 0.9668, Validation Accuracy: 0.9698, Loss: 0.0401
Epoch   3 Batch  410/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9570, Loss: 0.0395
Epoch   3 Batch  420/1077 - Train Accuracy: 0.9734, Validati

Epoch   4 Batch  130/1077 - Train Accuracy: 0.9710, Validation Accuracy: 0.9641, Loss: 0.0224
Epoch   4 Batch  140/1077 - Train Accuracy: 0.9688, Validation Accuracy: 0.9627, Loss: 0.0211
Epoch   4 Batch  150/1077 - Train Accuracy: 0.9710, Validation Accuracy: 0.9723, Loss: 0.0195
Epoch   4 Batch  160/1077 - Train Accuracy: 0.9711, Validation Accuracy: 0.9759, Loss: 0.0199
Epoch   4 Batch  170/1077 - Train Accuracy: 0.9621, Validation Accuracy: 0.9698, Loss: 0.0265
Epoch   4 Batch  180/1077 - Train Accuracy: 0.9703, Validation Accuracy: 0.9638, Loss: 0.0164
Epoch   4 Batch  190/1077 - Train Accuracy: 0.9789, Validation Accuracy: 0.9588, Loss: 0.0180
Epoch   4 Batch  200/1077 - Train Accuracy: 0.9762, Validation Accuracy: 0.9663, Loss: 0.0236
Epoch   4 Batch  210/1077 - Train Accuracy: 0.9635, Validation Accuracy: 0.9670, Loss: 0.0243
Epoch   4 Batch  220/1077 - Train Accuracy: 0.9626, Validation Accuracy: 0.9585, Loss: 0.0282
Epoch   4 Batch  230/1077 - Train Accuracy: 0.9624, Validati

Epoch   4 Batch 1010/1077 - Train Accuracy: 0.9887, Validation Accuracy: 0.9599, Loss: 0.0142
Epoch   4 Batch 1020/1077 - Train Accuracy: 0.9930, Validation Accuracy: 0.9716, Loss: 0.0146
Epoch   4 Batch 1030/1077 - Train Accuracy: 0.9910, Validation Accuracy: 0.9766, Loss: 0.0188
Epoch   4 Batch 1040/1077 - Train Accuracy: 0.9593, Validation Accuracy: 0.9744, Loss: 0.0248
Epoch   4 Batch 1050/1077 - Train Accuracy: 0.9820, Validation Accuracy: 0.9716, Loss: 0.0119
Epoch   4 Batch 1060/1077 - Train Accuracy: 0.9812, Validation Accuracy: 0.9666, Loss: 0.0149
Epoch   4 Batch 1070/1077 - Train Accuracy: 0.9531, Validation Accuracy: 0.9741, Loss: 0.0150
Epoch   5 Batch   10/1077 - Train Accuracy: 0.9642, Validation Accuracy: 0.9613, Loss: 0.0249
Epoch   5 Batch   20/1077 - Train Accuracy: 0.9836, Validation Accuracy: 0.9688, Loss: 0.0182
Epoch   5 Batch   30/1077 - Train Accuracy: 0.9863, Validation Accuracy: 0.9712, Loss: 0.0169
Epoch   5 Batch   40/1077 - Train Accuracy: 0.9852, Validati

Epoch   5 Batch  820/1077 - Train Accuracy: 0.9840, Validation Accuracy: 0.9766, Loss: 0.0119
Epoch   5 Batch  830/1077 - Train Accuracy: 0.9676, Validation Accuracy: 0.9670, Loss: 0.0200
Epoch   5 Batch  840/1077 - Train Accuracy: 0.9875, Validation Accuracy: 0.9691, Loss: 0.0143
Epoch   5 Batch  850/1077 - Train Accuracy: 0.9747, Validation Accuracy: 0.9766, Loss: 0.0275
Epoch   5 Batch  860/1077 - Train Accuracy: 0.9795, Validation Accuracy: 0.9751, Loss: 0.0147
Epoch   5 Batch  870/1077 - Train Accuracy: 0.9774, Validation Accuracy: 0.9638, Loss: 0.0163
Epoch   5 Batch  880/1077 - Train Accuracy: 0.9914, Validation Accuracy: 0.9574, Loss: 0.0202
Epoch   5 Batch  890/1077 - Train Accuracy: 0.9714, Validation Accuracy: 0.9716, Loss: 0.0167
Epoch   5 Batch  900/1077 - Train Accuracy: 0.9809, Validation Accuracy: 0.9645, Loss: 0.0187
Epoch   5 Batch  910/1077 - Train Accuracy: 0.9695, Validation Accuracy: 0.9716, Loss: 0.0207
Epoch   5 Batch  920/1077 - Train Accuracy: 0.9773, Validati

Epoch   6 Batch  630/1077 - Train Accuracy: 0.9922, Validation Accuracy: 0.9787, Loss: 0.0135
Epoch   6 Batch  640/1077 - Train Accuracy: 0.9702, Validation Accuracy: 0.9833, Loss: 0.0134
Epoch   6 Batch  650/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9815, Loss: 0.0112
Epoch   6 Batch  660/1077 - Train Accuracy: 0.9949, Validation Accuracy: 0.9837, Loss: 0.0094
Epoch   6 Batch  670/1077 - Train Accuracy: 0.9933, Validation Accuracy: 0.9744, Loss: 0.0138
Epoch   6 Batch  680/1077 - Train Accuracy: 0.9669, Validation Accuracy: 0.9741, Loss: 0.0158
Epoch   6 Batch  690/1077 - Train Accuracy: 0.9703, Validation Accuracy: 0.9695, Loss: 0.0145
Epoch   6 Batch  700/1077 - Train Accuracy: 0.9855, Validation Accuracy: 0.9808, Loss: 0.0139
Epoch   6 Batch  710/1077 - Train Accuracy: 0.9875, Validation Accuracy: 0.9769, Loss: 0.0148
Epoch   6 Batch  720/1077 - Train Accuracy: 0.9844, Validation Accuracy: 0.9737, Loss: 0.0147
Epoch   6 Batch  730/1077 - Train Accuracy: 0.9770, Validati

Epoch   7 Batch  440/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9517, Loss: 0.0131
Epoch   7 Batch  450/1077 - Train Accuracy: 0.9801, Validation Accuracy: 0.9634, Loss: 0.0176
Epoch   7 Batch  460/1077 - Train Accuracy: 0.9887, Validation Accuracy: 0.9638, Loss: 0.0117
Epoch   7 Batch  470/1077 - Train Accuracy: 0.9856, Validation Accuracy: 0.9787, Loss: 0.0124
Epoch   7 Batch  480/1077 - Train Accuracy: 0.9868, Validation Accuracy: 0.9666, Loss: 0.0139
Epoch   7 Batch  490/1077 - Train Accuracy: 0.9785, Validation Accuracy: 0.9691, Loss: 0.0125
Epoch   7 Batch  500/1077 - Train Accuracy: 0.9941, Validation Accuracy: 0.9592, Loss: 0.0091
Epoch   7 Batch  510/1077 - Train Accuracy: 0.9715, Validation Accuracy: 0.9638, Loss: 0.0118
Epoch   7 Batch  520/1077 - Train Accuracy: 0.9974, Validation Accuracy: 0.9744, Loss: 0.0090
Epoch   7 Batch  530/1077 - Train Accuracy: 0.9684, Validation Accuracy: 0.9638, Loss: 0.0169
Epoch   7 Batch  540/1077 - Train Accuracy: 0.9980, Validati

Epoch   8 Batch  250/1077 - Train Accuracy: 0.9826, Validation Accuracy: 0.9744, Loss: 0.0162
Epoch   8 Batch  260/1077 - Train Accuracy: 0.9937, Validation Accuracy: 0.9723, Loss: 0.0070
Epoch   8 Batch  270/1077 - Train Accuracy: 0.9742, Validation Accuracy: 0.9805, Loss: 0.0149
Epoch   8 Batch  280/1077 - Train Accuracy: 0.9773, Validation Accuracy: 0.9705, Loss: 0.0147
Epoch   8 Batch  290/1077 - Train Accuracy: 0.9754, Validation Accuracy: 0.9879, Loss: 0.0166
Epoch   8 Batch  300/1077 - Train Accuracy: 0.9885, Validation Accuracy: 0.9826, Loss: 0.0069
Epoch   8 Batch  310/1077 - Train Accuracy: 0.9953, Validation Accuracy: 0.9790, Loss: 0.0070
Epoch   8 Batch  320/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9837, Loss: 0.0147
Epoch   8 Batch  330/1077 - Train Accuracy: 0.9703, Validation Accuracy: 0.9709, Loss: 0.0141
Epoch   8 Batch  340/1077 - Train Accuracy: 0.9938, Validation Accuracy: 0.9837, Loss: 0.0123
Epoch   8 Batch  350/1077 - Train Accuracy: 0.9820, Validati

Epoch   9 Batch   60/1077 - Train Accuracy: 0.9754, Validation Accuracy: 0.9719, Loss: 0.0076
Epoch   9 Batch   70/1077 - Train Accuracy: 0.9848, Validation Accuracy: 0.9776, Loss: 0.0115
Epoch   9 Batch   80/1077 - Train Accuracy: 0.9812, Validation Accuracy: 0.9801, Loss: 0.0101
Epoch   9 Batch   90/1077 - Train Accuracy: 0.9758, Validation Accuracy: 0.9801, Loss: 0.0078
Epoch   9 Batch  100/1077 - Train Accuracy: 0.9918, Validation Accuracy: 0.9780, Loss: 0.0084
Epoch   9 Batch  110/1077 - Train Accuracy: 0.9859, Validation Accuracy: 0.9826, Loss: 0.0082
Epoch   9 Batch  120/1077 - Train Accuracy: 0.9949, Validation Accuracy: 0.9801, Loss: 0.0123
Epoch   9 Batch  130/1077 - Train Accuracy: 0.9903, Validation Accuracy: 0.9780, Loss: 0.0071
Epoch   9 Batch  140/1077 - Train Accuracy: 0.9885, Validation Accuracy: 0.9702, Loss: 0.0142
Epoch   9 Batch  150/1077 - Train Accuracy: 0.9918, Validation Accuracy: 0.9702, Loss: 0.0076
Epoch   9 Batch  160/1077 - Train Accuracy: 0.9965, Validati

Epoch   9 Batch  940/1077 - Train Accuracy: 0.9668, Validation Accuracy: 0.9780, Loss: 0.0074
Epoch   9 Batch  950/1077 - Train Accuracy: 0.9985, Validation Accuracy: 0.9876, Loss: 0.0092
Epoch   9 Batch  960/1077 - Train Accuracy: 0.9896, Validation Accuracy: 0.9780, Loss: 0.0109
Epoch   9 Batch  970/1077 - Train Accuracy: 0.9832, Validation Accuracy: 0.9680, Loss: 0.0095
Epoch   9 Batch  980/1077 - Train Accuracy: 0.9820, Validation Accuracy: 0.9805, Loss: 0.0093
Epoch   9 Batch  990/1077 - Train Accuracy: 0.9889, Validation Accuracy: 0.9762, Loss: 0.0075
Epoch   9 Batch 1000/1077 - Train Accuracy: 0.9758, Validation Accuracy: 0.9737, Loss: 0.0071
Epoch   9 Batch 1010/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9751, Loss: 0.0098
Epoch   9 Batch 1020/1077 - Train Accuracy: 0.9934, Validation Accuracy: 0.9801, Loss: 0.0106
Epoch   9 Batch 1030/1077 - Train Accuracy: 0.9941, Validation Accuracy: 0.9851, Loss: 0.0060
Epoch   9 Batch 1040/1077 - Train Accuracy: 0.9807, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [49]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [50]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [51]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    
    lower_sent = sentence.lower()
    sent_id = list()
    for word in lower_sent.split(' '):
        try:
            sent_id.append(vocab_to_int[word])
        except:
            sent_id.append(vocab_to_int['<UNK>'])

    return sent_id


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [57]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [19, 211, 145, 7, 20, 150, 141]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [36, 256, 191, 281, 197, 112, 181, 147, 1]
  French Words: il a vu un vieux camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.